### PR TITLE
Add support for dynamic data from UI plugins to be included in generated config.js

### PR DESCRIPTION
**Pull request by waab76**
## Description
Some plugins have dynamic UI settings data that needs to be provided to the UI at startup time via inclusion in the generated config.js file.  This change adds a new `PluginUISettingsProvider` interface and a method for binding instances of that interface such that the data they provide will be included when the config.js file is generated at server startup.

This is a re-do of #10063 

## How Has This Been Tested?
This change has been tested locally using an in-development Enterprise plugin.

## Screenshots (if appropriate):
Screen Shot 2021-02-16 at 4 26 54 PM

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/web/resources/AppConfigResource.java [hunk 1]
```diff
@@ -68,7 +78,19 @@ public String get(@Context HttpHeaders headers) {
...
+            "pluginUISettings", buildPluginUISettings());
         return templateEngine.transform(template, model);
     }
+
+    private String buildPluginUISettings() {
+        Map<String, Object> pluginUISettings = settingsProviders.entrySet().stream().collect(Collectors.toMap(
+                Map.Entry::getKey,
+                entry -> entry.getValue().pluginSettings()));
+        try {
+            return objectMapper.writeValueAsString(pluginUISettings);
```
Comment (dennisoelkers) [inline graylog2-server/src/main/java/org/graylog2/web/resources/AppConfigResource.java:91]: For debugging purposes, it would be nice to pretty-print this. We can do it using `ObjectWriter#withPrettyPrinter`.

**Review by dennisoelkers — CHANGES_REQUESTED** [review]
Looks good so far, just one tiny nitpick!

File: graylog2-server/src/main/java/org/graylog2/web/resources/AppConfigResource.java (hunk 1 above)
Comment (waab76) [reply graylog2-server/src/main/java/org/graylog2/web/resources/AppConfigResource.java:91]: Here's what `config.js` looks like with the update in place:

Screen Shot 2021-03-01 at 11 33 24 AM

I'll push a new commit shortly

File: graylog2-server/src/main/java/org/graylog2/plugin/inject/Graylog2Module.java [hunk 2]
```diff
@@ -52,6 +52,7 @@
 import org.graylog2.plugin.outputs.MessageOutput;
 import org.graylog2.plugin.security.PasswordAlgorithm;
 import org.graylog2.plugin.security.PluginPermissions;
+import org.graylog2.web.PluginUISettingsProvider;
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/plugin/inject/Graylog2Module.java:55]: Unrelated change?

File: graylog2-server/src/main/java/org/graylog2/plugin/PluginModule.java [hunk 3]
```diff
@@ -323,4 +324,13 @@ protected void addAuthServiceBackend(String name,
         authServiceBackendBinder().addBinding(name).to(factoryClass);
         registerJacksonSubtype(configClass, name);
     }
+
+    protected MapBinder<String, PluginUISettingsProvider> pluginUISettingsProviderBinder() {
+        return MapBinder.newMapBinder(binder(), String.class, PluginUISettingsProvider.class);
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/plugin/PluginModule.java:329]: We need to make sure the map binder exists even when there is no plugin that's calling `addPluginUISettingsProvider()`. Otherwise the server will not start.

Similar to the following. We can initialize the settings map binder there as well.

https://github.com/Graylog2/graylog2-server/blob/e8dd88cf71296c13c30b5ba4de6f559ca9500a21/graylog2-server/src/main/java/org/graylog2/shared/bindings/PluginBindings.java#L42-L46

**Review by bernd — CHANGES_REQUESTED** [review]
Looks good! Only two comments:

File: graylog2-server/src/main/java/org/graylog2/plugin/PluginModule.java (hunk 3 above)
Comment (waab76) [reply graylog2-server/src/main/java/org/graylog2/plugin/PluginModule.java:329]: Will add.

File: graylog2-server/src/main/java/org/graylog2/plugin/inject/Graylog2Module.java (hunk 2 above)
Comment (waab76) [reply graylog2-server/src/main/java/org/graylog2/plugin/inject/Graylog2Module.java:55]: Yes, that was a leftover from a previous version.  Will clean up.

**Review by bernd — APPROVED** [review]
Looks good from my point of view. :smile: :+1: Thank you!

**Review by kyleknighted — APPROVED** [review]
implementation is working as expected!

**Review by dennisoelkers — APPROVED** [review]
Looks good, works great! :tada:
